test_conformance/extensions/cl_khr_command_buffer/basic_command_buffer.cpp

```diff
@@ -13,157 +13,146 @@
-        test_error(error, "clSetKernelArg failed");
-
-        if (simultaneous_use)
-        {
-            cl_command_buffer_properties_khr properties[3] = {
-                CL_COMMAND_BUFFER_FLAGS_KHR,
-                CL_COMMAND_BUFFER_SIMULTANEOUS_USE_KHR, 0
-            };
-            command_buffer =
-                clCreateCommandBufferKHR(1, &queue, properties, &error);
-        }
-        else
-        {
-            command_buffer =
-                clCreateCommandBufferKHR(1, &queue, nullptr, &error);
-        }
-        test_error(error, "clCreateCommandBufferKHR failed");
-
-        return CL_SUCCESS;
+    test_error(error,
+               "Unable to query CL_DEVICE_COMMAND_BUFFER_CAPABILITIES_KHR");
+    simultaneous_use_support = simultaneous_use_requested
+        && (capabilities & CL_COMMAND_BUFFER_CAPABILITY_SIMULTANEOUS_USE_KHR)
+            != 0;
+    out_of_order_support =
+        capabilities & CL_COMMAND_BUFFER_CAPABILITY_OUT_OF_ORDER_KHR;
+
+    if (elements <= 0)
+    {
+        return CL_INVALID_VALUE;
     }
+    num_elements = static_cast<size_t>(elements);
+
+    // Kernel performs a parallel copy from an input buffer to output buffer
+    // is created.
+    const char *kernel_str =
+        R"(
+    __kernel void copy(__global int* in, __global int* out, __global int* offset) {
+        size_t id = get_global_id(0);
+        int ind = offset[0] + id;
+        out[ind] = in[ind];
+    })";
+
+    error = create_single_kernel_helper_create_program(context, &program, 1,
+                                                       &kernel_str);
+    test_error(error, "Failed to create program with source");
+
+    error = clBuildProgram(program, 1, &device, nullptr, nullptr, nullptr);
+    test_error(error, "Failed to build program");
+
+    in_mem =
+        clCreateBuffer(context, CL_MEM_READ_ONLY,
+                       sizeof(cl_int) * num_elements * buffer_size_multiplier,
+                       nullptr, &error);
+    test_error(error, "clCreateBuffer failed");
+
+    out_mem =
+        clCreateBuffer(context, CL_MEM_WRITE_ONLY,
+                       sizeof(cl_int) * num_elements * buffer_size_multiplier,
+                       nullptr, &error);
+    test_error(error, "clCreateBuffer failed");
+
+    cl_int offset = 0;
+    off_mem = clCreateBuffer(context, CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR,
+                             sizeof(cl_int), &offset, &error);
+    test_error(error, "clCreateBuffer failed");
+
+    kernel = clCreateKernel(program, "copy", &error);
+    test_error(error, "Failed to create copy kernel");
+
+    error = clSetKernelArg(kernel, 0, sizeof(in_mem), &in_mem);
+    test_error(error, "clSetKernelArg failed");
+
+    error = clSetKernelArg(kernel, 1, sizeof(out_mem), &out_mem);
+    test_error(error, "clSetKernelArg failed");
+
+    error = clSetKernelArg(kernel, 2, sizeof(off_mem), &off_mem);
+    test_error(error, "clSetKernelArg failed");
+
+    if (simultaneous_use_support)
```

Comment:
It's problematic that you try to create a command buffer before checking CL_DEVICE_COMMAND_BUFFER_REQUIRED_QUEUE_PROPERTIES_KHR in the Skip() method. It should be the other way around.

Reply:
Looks like a good point to me. If queue will not meet the minimum properties specified by `CL_DEVICE_COMMAND_BUFFER_REQUIRED_QUEUE_PROPERTIES_KHR` then `clCreateCommandBufferKHR` will likely fail (`CL_INCOMPATIBLE_COMMAND_QUEUE_KHR`) instead of skipping the test.

My proposition is to move all `clGetDeviceInfo` queries to `BasicCommandBufferTest::Skip` method and change the order of `Skip` and `SetUp` calls in `MakeAndRunTest` function. I verified that approach with previous and new tests introduced in this PR - it needs some additional refactoring (and reviewing all remaining `cl_khr_command_buffer` related PRs) but it looks promising.

Please let me know if you have other propositions on your mind, thanks.

Reply:
No, this approach looks good to me.